thinc/layers/list2ragged.py

```diff
@@ -20,7 +20,7 @@ def list2ragged() -> Model[InT, OutT]:
 
 def forward(model: Model[InT, OutT], Xs: InT, is_train: bool) -> Tuple[OutT, Callable]:
     def backprop(dYr: OutT) -> InT:
-        return model.ops.unflatten(dYr.data, dYr.lengths)  # type: ignore
+        return cast(InT, model.ops.unflatten(dYr.data, dYr.lengths))
 
     lengths = model.ops.asarray1i([len(x) for x in Xs])
-    return Ragged(model.ops.flatten(Xs), lengths), backprop
+    return Ragged(model.ops.flatten(cast(List[Array2d], Xs)), lengths), backprop
```

Comment:
The fact that we need to cast to `List[Array2d]` here signals to me that something seems wrong upstream. Looking at `ops.flatten`, this is typed with `X: Sequence[ArrayT]` with `ArrayT = TypeVar("ArrayT", bound=ArrayXd)`. That forces us to cast to this list of union types that we've been aiming to avoid :(

Taking a step back, currently `types.py` has:
```
ListXd = Union[List["FloatsXd"], List["IntsXd"]]
```
I think this can't work and should be 
```
ListXd = Union[List["Floats1d"], List["Ints1d"], List["Floats2d"], List["Ints2d"], List["Floats3d"], List["Ints3d"], List["Floats4d"], List["Ints4d"]]
```
to ensure that each list is in fact of one internal type only. With that definition, we could type the `X` of `ops.flatten` as `ListXd`.

That will cause further `mypy` errors, but it looks like most of those make sense to change, e.g. L63 at `with_ragged` is now 
```
list2d_seq = cast(List[Array2d], seq)
``` 
and would have to become
```
list2d_seq = cast(List2d, seq)
```
which is more correct.

Reply:
Yes, the typing of `flatten` and `unflatten` in the PR prior to this review comment was problematic in that it declared the mixed-type-list idiom, meaning it was typically necessary to cast whatever went into or out of the methods. The reason I had chosen to declare the types like this was that I found it important to maintain the contained type within an input list as the output type (for `flatten`) / to maintain the input type as the contained type within the output list (for `unflatten`), i.e. to make it clear that if you put say an `Ints3d` in, you will always get a `List[Ints3d]` out.

I think I've now found a solution that offers the best of both worlds. which is essentially to use the type signatures @svlandeg suggested but additionally to use `@overload` signature declarations to express the bindings between specific array types and the corresponding array-containing list types. I have only added these for those types that seem to be used frequently: `Floats2d`; `Ints1d` and `Array2d`. If people think it sensible I could add more, but remembering that people can always use the `ListXd` / `ArrayXd` solution, casting if necessary, this did not seem necessary.

I am very confident now that this is an appropriate solution because far from this causing new `mypy` errors it rather allowed the removal of a slew of `cast`s.

Reply:
Yes - I think this is the perfect solution @richardpaulhudson !